src/lib/optimizer/strategy/chunk_pruning_rule.hpp

```diff
@@ -20,27 +20,36 @@ class PredicateNode;
 class Table;
 
 /**
- * This rule determines which chunks can be pruned from table scans based on
- * the predicates present in the LQP and stores that information in the stored
- * table nodes.
+ * For a given LQP, this rule determines the set of chunks that can be pruned from a table.
+ * To calculate the sets of pruned chunks, it analyzes the predicates of PredicateNodes in a given LQP.
+ * The resulting pruning information is stored inside the StoredTableNode objects.
  */
 class ChunkPruningRule : public AbstractRule {
  protected:
   void _apply_to_plan_without_subqueries(const std::shared_ptr<AbstractLQPNode>& lqp_root) const override;
 
-  static std::set<ChunkID> _compute_exclude_list(const Table& table, const AbstractExpression& predicate,
-                                                 const std::shared_ptr<StoredTableNode>& stored_table_node);
+  std::set<ChunkID> _compute_exclude_list(const Table& table,
+                                          const std::vector<std::shared_ptr<PredicateNode>>& predicate_chain,
+                                          const std::shared_ptr<StoredTableNode>& stored_table_node) const;
 
   // Check whether any of the statistics objects available for this Segment identify the predicate as prunable
   static bool _can_prune(const BaseAttributeStatistics& base_segment_statistics,
                          const PredicateCondition predicate_condition, const AllTypeVariant& variant_value,
                          const std::optional<AllTypeVariant>& variant_value2);
 
-  static bool _is_non_filtering_node(const AbstractLQPNode& node);
-
   static std::shared_ptr<TableStatistics> _prune_table_statistics(const TableStatistics& old_statistics,
                                                                   OperatorScanPredicate predicate,
                                                                   size_t num_rows_pruned);
+
+  static std::set<ChunkID> _intersect_chunk_ids(const std::vector<std::set<ChunkID>>& chunk_id_sets);
+
+  static std::vector<std::vector<std::shared_ptr<PredicateNode>>> _find_predicate_chains_recursively(
+      const std::shared_ptr<StoredTableNode>& stored_table_node, const std::shared_ptr<AbstractLQPNode>& node,
+      std::vector<std::shared_ptr<PredicateNode>> current_predicate_chain = {});
+
+ private:
+  // Caches intermediate results
+  mutable std::unordered_map<std::shared_ptr<PredicateNode>, std::set<ChunkID>> _excluded_chunk_ids_by_predicate_node;
```

Comment:
Why does this have to be mutable?

Reply:
```
/**
   * Caches intermediate results.
   * Mutable because it needs to be called from the _apply_to_plan_without_subqueries function, which is const.
   */
```